### ISSUE TITLE
[BUG] Jitting with JAX produces incorrect results with parameter broadcasting

### ISSUE THREAD

**Issue by eddddddy** [PennyLaneAI/pennylane#2762]
### Expected behavior

Consider the following QNode:
```python
dev = qml.device("default.mixed", wires=2)

@qml.qnode(dev, diff_method="backprop", interface="jax")
def circuit(x):
    qml.RX(x, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(1))
```
When the QNode is called with a parameter, the following should be output:
```pycon
>>> x = jnp.array([0.3, 0.6])
>>> jax.jit(circuit)(x)
[0.9553365 0.8253356]
```

### Actual behavior

Instead, the following is output:
```pycon
>>> x = jnp.array([0.3, 0.6])
>>> jax.jit(circuit)(x)
[0.95533645 0.95533645]
```
It seems that only the first parameter is used for all calculations and the rest are ignored.

### Additional information

If `diff_method="parameter-shift"` is set in the QNode, the issue disappears.

If `jax.jit` is removed in the above code, the issue disappears.

If `"default.qubit"` is used instead, the issue disappears.

The issue also appears if the QNode is decorated with `qml.transforms.batch_params`.

### System information

```shell
Name: PennyLane
Version: 0.24.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: c:\users\edward.jiang\documents\pennylane
Requires: numpy, scipy, networkx, retworkx, autograd, toml, appdirs, semantic-version, autoray, cachetools, pennylane-lightning
Required-by: PennyLane-Lightning

Platform info:           Windows-10-10.0.19042-SP0
Python version:          3.8.10
Numpy version:           1.22.3
Scipy version:           1.8.0
Installed devices:
- default.gaussian (PennyLane-0.24.0.dev0)
- default.mixed (PennyLane-0.24.0.dev0)
- default.mixed.autograd (PennyLane-0.24.0.dev0)
- default.qubit (PennyLane-0.24.0.dev0)
- default.qubit.autograd (PennyLane-0.24.0.dev0)
- default.qubit.jax (PennyLane-0.24.0.dev0)
- default.qubit.tf (PennyLane-0.24.0.dev0)
- default.qubit.torch (PennyLane-0.24.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.23.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by eddddddy** [comment]
Perhaps this is more related to the recent additions to JAX-JIT support than to parameter broadcasting? Not entirely sure 😕.

**Comment by antalszava** [comment]
Hi @eddddddy,

> Perhaps this is more related to the recent additions to JAX-JIT support than to parameter broadcasting? Not entirely sure confused.

as we noted this during our chat, the latest changes were added to the JAX-JIT interface and those changes are not should not be relevant for `diff_method="backprop"` (unless the interface is mistakenly being applied).

The idea behind interfaces is that
1) they convert between NumPy and ML framework (i.e., JAX) objects to execute circuits on any device and
2) they provide custom gradient rules.

Interfaces allow us to feed in ML objects into a QNode even when the target device would not natively understand the ML objects.

When we have `diff_method="backprop"`, none of those are required because the entire computation is performed using the ML framework's objects (e.g., `jnp.array` objects). In such a case with the `default.qubit` device, under the hood, we would be swapping our device to `default.qubit.jax`.

**Comment by eddddddy** [comment]
Referencing [this test](https://github.com/PennyLaneAI/pennylane/pull/2754#discussion_r903101424) here so we remember to change it once this is fixed.

**Comment by dwierichs** [comment]
Thanks for digging this one out! :+1:  The reason seems to be, that `qml.transforms.broadcast_expand` is not JIT compatible.

**Comment by dwierichs** [comment]
I looked into this once more.
The problem is that `qml.tape.QuantumTape.hash` is not jit-compatible. Replacing `tape.hash` by `hash(tape)` in the `cache_execute` pipeline resolves the issue as far as I can tell (note that `default.qubit` does not longer produce the bug, because it now makes use of broadcasting. Replace it by `default.mixed` for reproducing/testing purposes).

**Comment by dwierichs** [comment]
`tape.hash` also is quite slow:
```py
dev = qml.device("default.mixed", wires=2)

@qml.qnode(dev, diff_method="backprop", interface="jax", cache=True)
def circuit(x):
    qml.RX(x, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(1))
x = jnp.array([0.3, 0.6])

circuit(x)
tape = circuit.tape
%timeit circuit.tape.hash
222 µs ± 33.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit hash(circuit.tape)
142 ns ± 2.98 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
(note the different units)
I don't think this is critical, because we do not use these hashes a lot (basically once per tape in an execution pipeline, as far as I am aware).

**Comment by dwierichs** [comment]
I can no longer reproduce this bug on master. @eddddddy, @josh146 , can you confirm?

**Comment by josh146** [comment]
@dwierichs if the minimal non-working example in the issue now works in `master`, I would say safe to close

**Comment by dwierichs** [comment]
Made the mistake of closing this, because it is fixed for `"default.qubit"`, although I had mentioned already that this will happen :facepalm: : 
>With the introduction of parameter broadcasting to DefaultQubit, this problem will be gone, but only for DefaultQubit devices. In particular, DefaultMixed and basically all other devices will still have this problem.

I reopen the issue and modify the reported example to use `"default.mixed"` instead.
